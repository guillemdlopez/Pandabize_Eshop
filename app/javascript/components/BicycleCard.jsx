import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { backgroundColor, isColor } from "../config/functions";

const BicycleCard = ({ id, name, description, photo, customizations }) => {
  const lowestPrice = function () {
    const prices = customizations.map((custom) => custom.price);
    return Math.min(...prices);
  };

  return (
    <Link to={`bicycle/${id}`}>
      <div className="bike-card" data-bike-id={id}>
        <img src={photo} className="bike-img" />
        <div className="bike-card-content">
          <p>
            <strong>{name}</strong>
          </p>
          <p className="text-black-50">
            {description.slice(0, 40)}
            {description.length > 30 ? "..." : ""}
          </p>
          <div className="d-flex mb-3">
            {customizations.map((custom) =>
              isColor(custom.value) ? (
                <div
                  className="color-value-card mr-1"
                  style={backgroundColor(custom.value)}
                  key={custom.id}
                ></div>
              ) : (
                ""
              )
            )}
          </div>
          {lowestPrice() !== Infinity ? (
            <div className="d-flex align-items-center">
              <p className="text-muted mb-0">from</p>
              <p className="ml-2 lowest-price mb-0">{lowestPrice()} â‚¬</p>
            </div>
          ) : (
            ""
          )}
        </div>
      </div>
    </Link>
  );
};

BicycleCard.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  photo: PropTypes.string.isRequired,
  customizations: PropTypes.array.isRequired,
};

export default BicycleCard;
