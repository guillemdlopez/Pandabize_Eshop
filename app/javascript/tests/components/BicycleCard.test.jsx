import React from "react";
import { mount } from "enzyme";
import BicycleCard from "../../components/BicycleCard";
import { userContext } from "../../context/userContext";
import { BrowserRouter as Router, Switch } from "react-router-dom";

describe("Testing the <BicycleCard /> component", () => {
  const bicycleProps = {
    id: 26,
    name: "Prettiest Bicycle",
    description: "This is the prettiest bicycle in the world. Buy it!",
    photo: "image.jpg",
    customizations: [
      {
        id: 26,
        name: "Wheel size",
        value: "15 inches",
        price: 9.23,
        bicycle_id: 26,
      },
      {
        id: 148,
        name: "Wheel size",
        value: "16 inches",
        price: 6.15,
        bicycle_id: 26,
      },
    ],
  };
  const wrapper = mount(
    <Router>
      <Switch>
        <userContext.Provider value={{ user: "Guillem", userId: 7 }}>
          <BicycleCard {...bicycleProps} />
        </userContext.Provider>
      </Switch>
    </Router>
  );
  test("should render it correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("the link should point to the bicycle show", () => {
    const { id: bicycleId } = bicycleProps;
    expect(wrapper.find("a").props().href).toEqual(`/bicycle/${bicycleId}`);
  });

  test("should display the img with the right url", () => {
    const { photo } = bicycleProps;
    expect(wrapper.find("img").prop("src")).toEqual(photo);
  });

  test("should contain the name of the bicycle", () => {
    const { name } = bicycleProps;
    expect(wrapper.find(".bicycle-name").text().trim()).toEqual(name);
  });

  test("should display the description of the bicycle", () => {
    const { description } = bicycleProps;
    expect(wrapper.find(".bicycle-description").text()).toEqual(
      `${description.slice(0, 40)}...`
    );
  });

  test("should display the lowest price", () => {
    const prices = bicycleProps.customizations.map((custom) => custom.price);
    const lowest = Math.min(...prices);

    expect(wrapper.find(".lowest-price").text()).toEqual(`${lowest} â‚¬`);
  });
});
