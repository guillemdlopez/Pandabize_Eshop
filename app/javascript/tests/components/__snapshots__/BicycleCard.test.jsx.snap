// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <BicycleCard /> component should render it correctly 1`] = `
<a
  href="/bicycle/26"
  onClick={[Function]}
>
  <div
    className="bike-card"
    data-bike-id={26}
  >
    <img
      className="bike-img"
      src="image.jpg"
    />
    <div
      className="bike-card-content"
    >
      <p
        className="bicycle-name"
      >
        <strong>
          Prettiest Bicycle
        </strong>
      </p>
      <p
        className="bicycle-description text-black-50"
      >
        This is the prettiest bicycle in the wor
        ...
      </p>
      <div
        className="d-flex mb-3"
      />
      <div
        className="d-flex align-items-center"
      >
        <p
          className="text-muted mb-0"
        >
          from
        </p>
        <p
          className="ml-2 lowest-price mb-0"
        >
          6.15
           â‚¬
        </p>
      </div>
    </div>
  </div>
</a>
`;
